fe/fe-core/src/main/java/com/starrocks/connector/HMSPartitionBasedRemoteInfoSource.java

```diff
@@ -0,0 +1,133 @@
+    public static final String HMS_PARTITIONS_LIST_FIES_ASYNC_GET = "HMS.PARTITIONS.LIST_FS_ASYNC.GET";
+    public static final String HMS_PARTITIONS_LIST_FILES_ASYNC_WAIT = "HMS.PARTITIONS.LIST_FS_ASYNC.WAIT";
+    public static final int HMS_PARTITION_BATCH_SIZE_MIN = 16;
+    public static final int HMS_PARTITION_BATCH_SIZE_MAX = 128;
+
+    private GetRemoteFilesParams params = null;
+    private Function<GetRemoteFilesParams, List<Partition>> fnGetPartitionValues = null;
+
+    private Function<Partition, RemoteFileInfo> fnGetRemoteFileInfo = null;
+
+    public HMSPartitionBasedRemoteInfoSource(Executor executor, GetRemoteFilesParams params,
+                                             Function<Partition, RemoteFileInfo> fnGetRemoteFileInfo,
+                                             Function<GetRemoteFilesParams, List<Partition>> fnGetPartitionValues) {
+        super(executor);
+        this.params = params;
+        this.fnGetPartitionValues = fnGetPartitionValues;
+        this.fnGetRemoteFileInfo = fnGetRemoteFileInfo;
+    }
+
+    @Override
+    public List<RemoteFileInfo> getOutputs(int maxSize) {
+        try (Timer ignored = Tracers.watchScope(Tracers.Module.EXTERNAL, HMS_PARTITIONS_LIST_FIES_ASYNC_GET)) {
+            return super.getOutputs(maxSize);
+        }
+    }
+
+    @Override
+    public boolean hasMoreOutput() {
+        // `hasMoreOutput` will trigger tasks to generate output.
+        try (Timer ignored = Tracers.watchScope(Tracers.Module.EXTERNAL, HMS_PARTITIONS_LIST_FILES_ASYNC_WAIT)) {
+            return super.hasMoreOutput();
+        }
+    }
+
+    @Override
+    public int computeOutputSize(RemoteFileInfo output) {
+        List<RemoteFileDesc> files = (output.getFiles());
+        int size = 1;
+        if (files != null) {
+            size = Math.max(size, files.size());
+        }
+        return size;
+    }
+
+    @Override
+    public RemoteFileInfo getOutput() {
+        List<RemoteFileInfo> res = getOutputs(1);
+        Preconditions.checkArgument(res.size() == 1);
+        return res.get(0);
+    }
+
+    class ListPartitionFilesTask implements AsyncTaskQueue.Task {
+        Partition partition;
+        Object attachment;
+
+        @Override
+        public List run() throws InterruptedException {
+            RemoteFileInfo remoteFileInfo = fnGetRemoteFileInfo.apply(partition);
+            remoteFileInfo.setAttachment(attachment);
+            return List.of(remoteFileInfo);
+        }
+    }
+
+    class GetPartitionValuesTask implements AsyncTaskQueue.Task<RemoteFileInfo> {
+        List<GetRemoteFilesParams> requests;
+        int usedIndex = 0;
+
+        @Override
+        public List<RemoteFileInfo> run() throws InterruptedException {
+            if (requests == null) {
+                requests = params.partitionExponentially(HMS_PARTITION_BATCH_SIZE_MIN, HMS_PARTITION_BATCH_SIZE_MAX);
```

Comment:
why get partiton exponentially? it could be better than the old way?

Reply:
what's the old way?

old way is to get all partitions. And that's a huge sync operation.

the new way is to split parttiions into groups: small group will take less time to get result.